src/pages/index.js

```diff
@@ -1,95 +1,245 @@
+api
+  .getInitialCards()
+  .then((cards) => {
+    cardsList.renderItems(cards);
+  })
+  .catch((err) => {
+    console.error(err);
+  });
+
 // // ! ||--------------------------------------------------------------------------------||
 // // ! ||                                 Event Handlers                                 ||
 // // ! ||--------------------------------------------------------------------------------||
 
+function handleDeleteClick(card) {
+  const id = card.getID();
+  deleteCardModal.open(() => {
+    deleteCardModal.setLoading();
+    api
+      .removeCard(id)
+      .then(() => {
+        card.handleDeleteCard();
+        deleteCardModal.close();
+      })
+      .catch((error) => {
+        console.error("Error removing card:", error);
+      })
+      .finally(() => {
+        deleteCardModal.resetButtonText();
+      });
+  });
+}
+
+function handleAddLike(card) {
+  const id = card.getID();
+  api
+    .addLike(id)
+    .then((updatedCardData) => {
+      card.setLiked(true);
+    })
+    .catch((err) => {
+      console.error(err);
+    });
+}
+
+function handleRemoveLike(card) {
+  const id = card.getID();
+  api
+    .removeLike(id)
+    .then((updatedCardData) => {
+      card.setLiked(false);
+    })
+    .catch((err) => {
+      console.error(err);
+    });
+}
 function handleAddCardFormSubmit(inputValues) {
-  console.log(this._modalForm);
   const name = inputValues.title;
   const link = inputValues.url;
-  const newCardElement = createCard({ name, link });
-  cardsList.prependItem(newCardElement);
-  addCardModal.close();
-  addCardFormValidator.resetValidation();
-  addCardModal.reset();
+  addCardModal.setLoading();
+  api
+    .addCard({ name, link })
+    .then((newCardData) => {
+      const newCardElement = createCard(newCardData);
+      cardsList.prependItem(newCardElement);
+      addCardModal.close();
+      addCardFormValidator.resetValidation();
+      addCardModal.reset();
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      addCardModal.resetButton();
+    });
+}
+
+function handleUpdateAvatarFormSubmit(inputValues) {
+  const link = inputValues.url;
+  updateAvatarModal.setLoading();
+  api
+    .changeProfilePicture(link)
+    .then((data) => {
+      userInfo.setAvatar(data.avatar);
+      updateAvatarModal.reset();
+      updateAvatarModal.close();
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      updateAvatarModal.resetButton();
+    });
+}
+
+function handleUpdateAvatar() {
+  const link = document.getElementById("avatar-image-url").value;
+  updateAvatarModal.open(() => {
+    api
+      .changeProfilePicture(link)
+      .then(() => {
+        deleteCardModal.close();
+      })
+      .catch((error) => {
+        console.error("Error removing card:", error);
+      });
+  });
 }
 
 function handleProfileEditSubmit(inputValues) {
   userInfo.setUserInfo(inputValues);
+  profileEditPopup.setLoading();
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

Reply:
Cool suggestion, I played with this for a while but actually prefer the original implementation.  The universal function cuts down some lines of code, but I think it's easier to understand if the logic for the handlers is all in one place instead of divided

Reply:
It's just for avoiding repetition of the code: `then`, `catch`, finally`, loading effects and so on